uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/SamlLoginIT.java

```diff
@@ -445,21 +444,11 @@ public void testSingleLogoutWithNoLogoutUrlOnIDP() throws Exception {
 
         provider = IntegrationTestUtils.createOrUpdateProvider(zoneAdminToken, baseUrl, provider);
 
-        webDriver.get(baseUrl + "/login");
-        Assert.assertEquals("Cloud Foundry", webDriver.getTitle());
-        webDriver.findElement(By.xpath("//a[text()='" + provider.getConfig().getLinkText() + "']")).click();
-        webDriver.findElement(By.xpath(SIMPLESAMLPHP_LOGIN_PROMPT_XPATH_EXPR));
-        webDriver.findElement(By.name("username")).clear();
-        webDriver.findElement(By.name("username")).sendKeys(testAccounts.getUserName());
-        webDriver.findElement(By.name("password")).sendKeys(testAccounts.getPassword());
-        webDriver.findElement(By.xpath("//input[@value='Login']")).click();
-        assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Where to"));
-
-        webDriver.findElement(By.cssSelector(".dropdown-trigger")).click();
-        webDriver.findElement(By.linkText("Sign Out")).click();
-        webDriver.findElement(By.xpath("//a[text()='" + provider.getConfig().getLinkText() + "']")).click();
-
-        assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Where to"));
+        LoginPage.go(webDriver, baseUrl)
+                .startSamlLogin()
+                .login(testAccounts.getUserName(), testAccounts.getPassword())
+                .logout()
+                .automaticSamlLogin();
```

Comment:
Does `.automaticSamlLogin();` whose definition is just
```
public HomePage automaticSamlLogin() {
        clickFirstSamlLoginLink();
        return new HomePage(driver);
    }
```
cover the assertion `assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Where to"));`?

Reply:
I see, it does [here](https://github.com/cloudfoundry/uaa/blob/9108bcca50b86e3591e0b8590b58c0e20be66c74/uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/pageObjects/HomePage.java#L17-L18).

Reply:
My feeling is that `automaticSamlLogin();` abstracts out too much details & a new reader wouldn't understand the basic test logic without clicking into `automaticSamlLogin()`. 

I'm thinking about what would be readable. Maybe something like this:

        LoginPage.go(webDriver, baseUrl)
                .startSamlLogin()
                .login(testAccounts.getUserName(), testAccounts.getPassword())
                .logout()
                .clickFirstSamlLoginLink() 
                .assertOnUAAHomePage()

Reply:
> Does `.automaticSamlLogin();` ...
> cover the assertion `assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Where to"));`?

Yes, as you found, every time we go to a page, when we construct the object for it, it starts by validating that we're on that page. The page objects, however, could be changed to do that in a different way. So if a test really wanted to make sure that text was on the page, a test (ideally only 1 test), should have code to explicitly test that. We would add a method to the page Page object to facilitate this. In fact, I think I'll do some refactoring to show how.

Note that the new check looks at all the text on the page and doesn't care if it's in an h1 block. That's intended to make the test easier to maintain by removing unnecessary details.

Reply:
Great! FYI, I further edited my pseudocode above to demonstrate my points.

Reply:
> My feeling is that `automaticSamlLogin();` abstracts out too much details & a new reader wouldn't understand the basic test logic without clicking into `automaticSamlLogin()`.

clickFirstSamlLoginLink() is a private method, and it doesn't support the fluent interface. The design I'm coalescing on is having the expected landing page indicated by the choice of navigation method. Let's see if we can improve the automaticSamlLogin() name. How about goToSamlLoginPageWhenAlreadyLoggedIn()?

Would you want the landing page included in this and every other method that navigates to a different page? Something like goToSamlLoginPageWhichRedirectsToHomePage()? I don't think I like that as much.

Reply:
My guess is that a new reader wouldn't understand 
```
        LoginPage.go(webDriver, baseUrl)
                .startSamlLogin()
                .login(testAccounts.getUserName(), testAccounts.getPassword())
                .logout()
                .goToSamlLoginPageWhenAlreadyLoggedIn();
```
either.

Reply:
My guess is that a new reader wouldn't understand 
```
        LoginPage.go(webDriver, baseUrl)
                .startSamlLogin()
                .login(testAccounts.getUserName(), testAccounts.getPassword())
                .logout()
                .goToSamlLoginPageWhenAlreadyLoggedIn();
```
either without clicking into all these functions. Maybe we can ask a fresh pair of eyes to read this. @bruce-ricard?

Reply:
I think it's ok if someone needs to look at the implementation of a method to understand why the navigation is going the way it is, if the page objects are well-named and possibly commented. These navigation methods can serve as documentation for how the site is expected to work. But let's see if we can minimize how often that would need to happen.

This example you wrote is interesting. Some of the navigation methods clarify where they're coming from (like logoutOnUAAHomePage()), which helps because there are many pages you can log out from, and at least two different logout pages. Following that design, however, would mean that instead of relying on inheriting the logout method from the base class like we do now, we would need to have a method in each class we need to log out from in order to give it that unique name.

Changing to using methods like assertOnSamlLoginPage() would result in more verbose tests, but it would perhaps make the flow easier to follow. I think what we'd need to do is that these validator also be the methods that return the new page objects for the new current page.

Reply:
I just pushed a refactor. I think it's clear now which page we're coming from, what action is taken, and where we're going. @peterhaochen47 what do you think?

```
        LoginPage.go(webDriver, baseUrl)
                .clickSamlLink_goToSamlLoginPage()
                .login_goToHomePage(testAccounts.getUserName(), testAccounts.getPassword())
                .logout_goToLoginPage()
                .clickSamlLink_goToSamlLoginPage();
```